### ISSUE TITLE
Feature Request - Customizable Feed System Inspired by Bluesky

### ISSUE THREAD

**Issue by T3kn3vr4** [mastodon/mastodon#33098]
### Pitch

Implementation Of A Customizable Feed Feature Similar To The One Currently Offered By Bluesky.

Key Features to Consider:

1. User-Created Feeds:
   Allow users to create and customize their own feeds based on specific topi

2. Flexible Feed Definition:
   Enable feed creation using various parameters such as hashtags, emojis, keywords, or user lists.

3. Content Filtering:
   Implement options to filter content by post type (e.g., images, videos, text-only posts).

4. Real-Time Updates:
   Ensure feeds capture and display new posts in real-time for a dynamic user experience.

5. Feed Management Tools:
   Introduce features like pinning important posts, blocking specific users from feeds, and setting up human review for sensitive content.

6. Discover Feed Personalization:
   Develop a system that allows users to refine their Discover feed with options like "Show more like this" or "Show less like this."

7. Open Feed Marketplace:
   Create a platform where users can share and discover custom feeds created by others in the community.

8. Developer Support:
   Provide APIs and tools for third-party developers to create and integrate custom feeds into the Mastodon ecosystem.

I personally believe that implementing this feature would not only enhance Mastodon's functionality but also maintain its competitive edge in the evolving social media landscape.

________________________________________________

ALGORITHMIC CHOICE WITH CUSTOM FEEDS:
https://bsky.social/about/blog/7-27-2023-custom-feeds

BLUESKY ROLLS OUT FEEDS WITH CUSTOM ALGORITHMS / YES, THERE IS AN ALF-FOCUSED FEED.:
https://www.theverge.com/2023/5/26/23739174/bluesky-custom-feeds-algorithms-twitter-alternative

### Motivation

Benefits:

- Increased User Engagement: Customizable feeds cater to individual interests, potentially increasing time spent on the platform.
- Community Building: Shared feeds can foster niche communities and connections among users with similar interests.
- Content Discovery: A more refined content discovery process can help users find relevant posts and accounts more easily.
- Reduced Algorithm Dependence: This feature would align with Mastodon's ethos of user control and decentralization by reducing reliance on opaque algorithmic recommendations.

**Comment by golfinq** [comment]
Inspired by https://github.com/mastodon/mastodon/issues/309#issuecomment-2507381505; I found this issue. I am not sure how it is implemented in bluesky under the hood but, I think one way to actually implement this feature would be using a simple programming language which would describe how to weight posts simliar to the [pandas framework query/eval](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.eval.html) syntax. It would ingest either the followers feed, a customized set of users, or an entire servers feed then run it through the custom weighting algorithm.

**Comment by T3kn3vr4** [comment]
Here's some more information that I was able to find online:



https://docs.bsky.app/docs/starter-templates/custom-feeds?utm_source=perplexity

https://docs.bsky.app/docs/tutorials/custom-feeds?utm_source=perplexity

https://github.com/MarshalX/bluesky-feed-generator?utm_source=perplexity

https://docs.bsky.app/docs/starter-templates/custom-feeds?utm_source=perplexity

https://markptorres.com/research/2023-11-29-working-with-bsky?utm_source=perplexity

**Comment by piotrek-k** [comment]
> I think one way to actually implement this feature would be using a simple programming language which would describe how to weight posts 

@golfinq Implementing custom programming language for that would be hard to do and at the same time very limiting. Only simple operations would be possible. If somehow that language was turing complete, you would still need to limit its throughput - w don't want to put too much load on servers, aren't we? Indeally, custom feeds should offer not only simple filtering but also ability to use more advanced ML algorithms to find interesting content.

How I see it:
1. Mastodon servers should offer a way to retrieve new posts in real time by external services
2. External services should filter received data on their own
3. External services should serve feeds through some endpoint (maybe event customized per user)
4. Finally, Mastodon servers should be able to call those services on behalf of users and serve them their feeds

This way we are not limiting what custom feeds can do and offload computation to external services. There are, though, some privacy issues to be addressed. It would be yet another service processing user activity data.

**Comment by michelcrypt4d4mus** [comment]
i've been working on a javascript package that implements this kind of feed algorithm (and let's you configure the decay and weighting etc): https://github.com/michelcrypt4d4mus/fedialgo_demo_app_foryoufeed

<img width="1504" alt="Image" src="https://github.com/user-attachments/assets/3eeca666-f76b-4ad0-9eca-914df547d209" />

**Comment by golfinq** [comment]
After consideration, I agree that we shouldn't use the back-end and implementing an API for external services and/or operating in the front-end is better.

**Comment by michelcrypt4d4mus** [comment]
yeah having built this out over the last few weeks i don't think there's any need to implement this server side. FediAlgo manages to accomplish everything with the available Mastodon API as is.

**Comment by golfinq** [comment]
@reiver@mastodon.social is discussing this at a protocol level with a new activitypub "feed" object - which seems like a good idea

https://mastodon.social/@reiver/113668493283013849

**Comment by T3kn3vr4** [comment]
@golfinq 

Has there been any more progress on the feed feature?

All that I'm seeing from the Mastodon link that you posted was from December 2024.

**Comment by michelcrypt4d4mus** [comment]
not sure what you mean? `fedialgo` is 100% usable as an npm library now and has a stable API that will not change for the foreseeable future.

**Comment by T3kn3vr4** [comment]
You should consider potentially expanding Fedialgo from just Mastodon to other Fediverse platforms.

I. e allow Lemmy users to create a sort of feed algorithm, or Friendica, or etc. 

Fedialgo could potentially be the home for Fediverse - style, user created/controlled Feeds.

**Comment by michelcrypt4d4mus** [comment]
i'm not a lemmy user (yet) so that's probably not particularly likely in the near term unless the lemmy API is basically the same as the mastodon API but at the same time the interactions w/mastodon are localized into just two places in the `fedialgo` codebase so it theoretically wouldn't be super hard to implement another backend for lemmy.

**Comment by brendanjones** [comment]
I completely agree that Mastodon needs a 'feeds' functionality that's as powerful (or near enough) as Bsky's. I think here on Mastodon we can do better: all feed and list functionality can be included in lists. I think Bluesky made a mistake, there's no need to separate the two.

Given that, let's look at this functionality being added to lists. Best for each feature to be a single issue, there's no way it's all getting added at once (nor does it need to be).

Lots of existing issues asking for the (what I think are the) most important features of feed-like lists:
- Share/follow lists: https://github.com/mastodon/mastodon/issues/23276 and https://github.com/mastodon/mastodon/issues/18292, with the sharing part maybe solved by https://github.com/mastodon/mastodon/pull/33121.
- Follow hashtags in lists: https://github.com/mastodon/mastodon/issues/7692
- Follow keywords in lists: not yet requested
- Filter keywords in lists: https://github.com/mastodon/mastodon/issues/21581 (also needs https://github.com/mastodon/mastodon/issues/21388 or https://github.com/mastodon/mastodon/issues/7953).
- Filter by post type (e.g., images, videos, text-only posts): https://github.com/mastodon/mastodon/issues/30104 (sort of, this issue is vague and not specific to lists)
- List marketplace: share and discover custom feeds created by others in the community: not yet requested
- Custom sorting: https://github.com/mastodon/mastodon/issues/3782 (though not specific to lists) and already existing in Phanpy for your home feed at least -try it out at https://phanpy.social/#/catchup .

So yeah, a lot of this is already requested. It requires a concerted and coordinated effort to implement as a proper 'feed' feature, though.

"Feed Management Tools" feels very non-MVP to me. "Discover Feed Personalization" I don't think is related, feeds/lists are not your discover feed, and in any case I doubt this kind of feature will be done.

**Comment by michelcrypt4d4mus** [comment]
@nanos was kind enough to deploy a publicly accessible version of the FediAlgo demo app so people can try it out here: [fedialgo.thms.uk/](https://fedialgo.thms.uk/)

@brendanjones from your list:

> * Filter keywords in lists
> * Filter by post type

The demo app can filter by post type and can filter for one or more hashtags (see screenshots)

<img width="790" alt="Image" src="https://github.com/user-attachments/assets/0ab625da-5503-4e42-a1f1-c0aee488369f" />

<img width="1620" alt="Image" src="https://github.com/user-attachments/assets/23d34b54-4bb8-44a5-b04b-dd304dd3d1bc" />
